### ISSUE TITLE
Merge `node` changes from @types org

### ISSUE THREAD

**Issue by blakeembrey** [DefinitelyTyped/DefinitelyTyped#14418]
There's a lot of commits on https://github.com/types/env-node that improve node compatibility that should be pulled into the node definition here to make it more useable. For instance, the `stream` module has the proper structure (`Duplex` is an instance of `Transform`, which is an instance of `Readable` that implements `Writeable`) while parts like `fs`, `util`, `crypto`, `module`, `http` have all been completely re-typed from the documentation (at least as of last year with 6.0). I know there's half a dozen issues with the current `node/index.d.ts` because it's unusable on any heavy node projects (I mostly run into the stream incompatibility issue because I can't assign a pass through stream to a readable stream, but there's a lot others for my more node-specific use-cases like `ts-node` and Express.js middleware). What's the best way to do this?

**Comment by mhegazy** [comment]
Is your concern the breaking changes? Or just the size of the work involved? If it is the later I would love to help.

**Comment by blakeembrey** [comment]
Bit of both. It's definitely a non-trivial load of work to merge them together and will likely have at least a few breaking changes. The breaking changes can probably be managed though since I don't think there's many incorrect breakages, but there's some other parts it brings up (like how long a TypeScript interface should be there incorrectly until it's deprecated).

What do you feel is the best way to start on the work? Should we try built-in module by built-in module via PRs? Most of the changes will need to be back-ported through all the supported node.js versions too.

Once redirects exist too, can we move the types into their own repo? I'd like to be able to continue maintaining them.

**Comment by winksaville** [comment]
I'm just starting with node and I did `npm install @types/node` and the latest version was 7.0.4:
```
$ npm install
test-simple-node-typescript@0.1.0 /home/wink/prgs/test-simple-node-typescript
├── @types/node@7.0.4 
└── typescript@2.1.5 
```

Where as the latest node is 7.5.0:
```
$ node --version
v7.5.0
```

What is the relationship between the version of node and that of @types/node?

**Comment by blakeembrey** [comment]
Commenting on a random issue isn't the best way to get answers you're looking for. It causes confusion and distracts from the thread, please avoid it. You should read the existing thread first (to verify if it's related to your issue) and open a new issue unless it appears you're having the same problem, in which case certainly leave a comment to avoid duplicate issues.

The answer to your question can be found in the README: https://github.com/DefinitelyTyped/DefinitelyTyped#what-exactly-is-the-relationship-between-this-repository-and-the-types-packages-on-npm.

**Comment by mhegazy** [comment]
> but there's some other parts it brings up (like how long a TypeScript interface should be there incorrectly until it's deprecated such as the case of ServerResponse - a class/interface that doesn't even exist in node but is in DefinitelyTyped because of legacy).

I do not think there has been a policy for this. also cc @vvakame, @RyanCavanaugh and @andy-ms. My gut feel is to first add the non-breaking changes, or the ones that are just wrong. then keep removing/depreicating the interfaces that should not be there later. 

> What do you feel is the best way to start on the work? Should we try built-in module by built-in module via PRs? Most of the changes will need to be back-ported through all the supported node.js versions too.

Sounds like a plan. Lets start and see how it goes.

**Comment by blakeembrey** [comment]
> then keep removing/depreicating the interfaces that should not be there later.

Not sure I agree, unless there's some tooling in place to notify later. Finding a way to do independent semver with definitions to the source would help here though, since I understand the concern as it'll break _everyone_. No amount of waiting can really solve that unfortunately. I believe the some incorrect interface has been there at least year, probably more. Hard to remove once added with no versioning.

One thought on making it more manage-able, what if the TypeScript compiler warned on deprecated functions (using `@deprecated` in JSDoc)? It's pretty standard and would help people trying to upgrade too (runtime warning for deprecation is common today, this would complement and be compile-time deprecation). Then the deprecation plan could reasonably be implemented as there could be a planned deprecation timeline.

> Sounds like a plan. Lets start and see how it goes.

Ok, it might take me a long time though with all my other commitments but anyone else is welcome to jump in and help 😄

**Comment by hinell** [comment]
>ServerResponse - a class/interface that doesn't even exist in node but is in DefinitelyTyped because of legacy).
Well, it definitely does :laughing: :
https://nodejs.org/dist/latest-v7.x/docs/api/http.html#http_class_http_serverresponse

**Comment by blakeembrey** [comment]
Sorry, that was a typo - meant to be ServerRequest. The only incompatibilities I've seen are caused by incompatible names and using deprecated interfaces. The node definition here has a lot more issues, such as basic things like Readable and PassThrough stream incompatibility.

**Comment by hinell** [comment]
>ServerRequest

Already [deprecated](../blob/master/node/index.d.ts#L641).

>such as basic things like Readable and PassThrough stream incompatibility.
Need more info

**Comment by blakeembrey** [comment]
Yes, I know that. I'm saying it's not removed. Nothing is stopping anyone using the interface. What more information could you need for fixing streams? I am saying it's broken right now because the transform, duplex and pass through streams **are not** assignable to Readable, which makes it impossible to use in any heavy node applications using streams.

**Comment by hinell** [comment]
Well, I think it's better to merge dt/node into type/node rather than vice versa.

**Comment by mhegazy** [comment]
> Well, I think it's better to merge dt/node into type/node rather than vice versa.

The effect on the user is the same in either cases.  there are going to better declarations for some modules, and breaking changes. does not really matter which way you merge it.

**Comment by hinell** [comment]
>does not really matter which way you merge it

Yep that doesn't but really matter how to maintain it & use it. 
[Here related thread](../../Microsoft/TypeScript/issues/11973)
@mhegazy you're already participated

**Comment by blakeembrey** [comment]
> Well, I think it's better to merge dt/node into type/node rather than vice versa.

Sure, I can definitely do that, I'm only suggesting the other direction because there's no timeline on redirects and it's causing a lot of issues with interop (mostly things like `Readable` are incompatible, and some server work like https://github.com/blakeembrey/node-servie can not run using DefinitelyTyped's version although I'm trying really hard to use NPM's `@types`). I would definitely prefer to keep node definitions in a separate repo (or as many as possible) so that I can help respond to issues and maintain, which today I can't in DefinitelyTyped, but I'm not sure if that's the point you're trying to make.

I can make an effort to pull everything from DefinitelyTyped into @types first, then go the other direction - it'll probably take me a day or so to review all the work done on DefinitelyTyped, which is obviously non-trivial when I'm working on all of @types for free 😄

**Comment by hinell** [comment]
>I would definitely prefer to keep node definitions in a separate repo (or as many as possible) so that I can help respond to issues and maintain, which today I can't in DefinitelyTyped

If so, you're welcome to open a new PR.  I even can help you to resolve the issues arising from types/node integration. You can add me as collaborator into your fork if you want. 

>I'm not sure if that's the point you're trying to make

I meant to merge dt/node into types/node with intent to bring DT-compatible interfaces into node, that would allow you to use it  as separate repo and have it compatible with all DT types.

**Comment by blakeembrey** [comment]
> If so, you're welcome to open a new PR. I even can help you to resolve the issues arising from types/node integration. You can add me as collaborator into your fork if you want.

I'm not sure what's being miscommunicated here, the entire intent was that I'd be making PRs to try to bring DefinitelyTyped up to speed with the changes I've done in an external repo. I don't see how that's related to adding you as a collaborator either.

> I meant to merge dt/node into types/node with intent to bring DT-compatible interfaces into node, that would allow you to use it as separate repo and have it compatible with all DT types.

I'm not sure you understand. TypeScript has no way to use definitions from a separate repo, if I want users to have a correct definition for node I need to contribute it back here eventually and that's what I'm trying to do. What exactly do you expect to achieve from the "DT-compatible" interface changes? I haven't seen any incompatibilities except where DefinitelyTyped was just wrong and I've fixed it in @types. If there's something that needs to improving, you should point it out - I definitely do not have the time to audit every single line of code, I intended to bring it back up to compatibility from when they forked which is only a couple of months worth of commits.

**Comment by hinell** [comment]
>I don't see how that's related to adding you as a collaborator either.

I just mean after you would have the `types/node` version being merged into your `dt/node` one I could help you with resolving test issues if they would be present.

>I need to contribute it back here eventually and that's what I'm trying to do.
> I haven't seen any incompatibilities except where DefinitelyTyped was just wrong and I've fixed it in @types.

Wait. Are you just asking us how to easily restyle the node declaration?
Just lost exact point what we are talking about. :confused:

Pls, clarify this excerpt from your issue description:
>What's the best way to do this?

**Comment by blakeembrey** [comment]
@hinell Please try re-reading the issue. It was already solved with @mhegazy before you joined in. If there's something unclear, please ask a question.

**Comment by blakeembrey** [comment]
> I just mean after you would have the types/node version being merged into your dt/node one I could help you with resolving test issues if they would be present.

I'm confused, sorry - do you want to be made a collaborator of @types or @DefinitelyTyped? Either way, I don't think it'll be necessary, I'll be pulling in the changes as I can or as someone else can look at them.

> Wait. Are you just asking us how to easily restyle the node declaration?

I don't really know what this means. What are you talking about?

> Pls, clarify this excerpt from your issue description:

See the title.

**Comment by hinell** [comment]
>do you want to be made a collaborator of @types or @DefinitelyTyped?

No I don't. Just would love to help you with breaking changes. 

>I don't really know what this means. What are you talking about?

Well,I think I'm a little bit confused. That's too long conversation. :laughing: 

>What's the best way to do this?

Eventually my answer is just to stop talking about and just open a new PR. 
As far as I know there is no best or easier way to merge external declarations into DT or a universal recipe exists.
Notify this issue after you have open the new pr.

**Comment by blakeembrey** [comment]
> Eventually my answer is just to stop talking about and just open a new PR.

This was my intention before you commented. This issue is to:

1. find out the preferred way to merge these types for the TypeScript team - just copying it outright is likely too much to review in a single go, plus there'll be a dozen breaking changes and likely some regressions
2. have an issue to track the merge publicly for anything that does come up, since this will likely be run over a number of iterations (e.g. how to handle deprecation, the obvious one being `ServerRequest` which has been deprecated forever while there's likely going to be new ones as I fix things)

Instead, I got into this conversation out of a typo that has caused a lot of confusion instead. The answers around handling deprecation are still unsolved and relevant - it's really up to the TypeScript team to decide how semver should be handled there. At some point, someone needs to bite the bullet and delete it, but how?

Since there's something you felt you needed to answer for me, can you let me know what you think the question was and then I can respond accordingly?

**Comment by hinell** [comment]
>This issue is to: find out the preferred way to merge these types for the TypeScript team

There is obviously only one way to achieve this: step by step, interface by interface. 
You don't want to do that otherwise. 

>what you think the question was and then I can respond accordingly?

Well initially I  misunderstood your question as of asking for moving dt/node repo into external one and asking community's opinion. 
Then I almost lost your point when refused myself to read all conversation. :no_mouth: 
Rereading has returned me back.

>Since there's something you felt you needed to answer for me

From now on I've no question you can answer on.

Don't forget to link this issue when will start PRing.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
